crates/sparrow-backend/src/exprs/expr_lang.rs

```diff
@@ -0,0 +1,68 @@
+use arrow_schema::DataType;
+use egg::Id;
+use smallvec::SmallVec;
+use sparrow_arrow::scalar_value::ScalarValue;
+
+use crate::Error;
+
+#[derive(Hash, PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
+pub(crate) struct ExprLang {
+    /// The name of the instruction being applied by this expression.
+    ///
+    /// Similar to an opcode or function.
+    ///
+    /// Generally, interning owned strings to the specific static strings is preferred.
+    pub name: &'static str,
+    /// Literal arguments to the expression.
+    pub literal_args: SmallVec<[ScalarValue; 2]>,
+    /// Arguments to the expression.
+    pub args: SmallVec<[egg::Id; 2]>,
+    // TODO: This includes the `DataType` in the enodes.
+    // This is necessary for ensuring that cast instructions to different types are treated
+    // as distinct, however it is potentially risky for writing simplifications, since the
+    // patterns won't have specific types. We may need to make this optional, so only the
+    // cast instruction has to specify it, and then rely on analysis to infer the types.
+    pub result_type: DataType,
+}
+
+// It is weird that we need to implement `Display` for `ExprLang` to pretty print
+// only the kind. But, this is a requirement of `egg`.
+impl std::fmt::Display for ExprLang {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.name.fmt(f)
+    }
+}
+
+impl egg::Language for ExprLang {
+    fn children(&self) -> &[egg::Id] {
+        &self.args
+    }
+
+    fn children_mut(&mut self) -> &mut [egg::Id] {
+        &mut self.args
+    }
+
+    fn matches(&self, other: &Self) -> bool {
+        // Note that this should *only* consider whether the node is the same operation.
+        //
+        // egg itself will check whether the arguments are *equivalent*.
+        self.name == other.name && self.literal_args == other.literal_args
```

Comment:
A) Should this also be comparing `args`? 
B) Should this also compare `result_type`? e.g. for `casts` that take the same input but cast to a different type?

Reply:
A) No -- see the comment here as well as docs for the trait (https://egraphs-good.github.io/egg/egg/trait.Language.html#tymethod.matches).
B) Hmm. Yes, for exactly that reason.